components/datetime/src/format/input.rs

```diff
@@ -17,24 +18,106 @@ use icu_time::{zone::UtcOffset, Time, TimeZone};
 // TODO(#2630) fix up imports to directly import from icu_calendar
 pub(crate) use icu_calendar::types::{DayOfMonth, MonthInfo, Weekday, YearInfo};
 
-#[derive(Debug, Copy, Clone)]
-pub(crate) struct ExtractedInput {
+#[cfg(doc)]
+use crate::input::*;
+
+/// An input bag with all possible datetime input fields.
+///
+/// Each input field may or may not be required, depending on the field set
+/// and the options.
+#[derive(Debug, Copy, Clone, Default)]
+#[non_exhaustive]
+pub struct DateTimeInputUnchecked {
+    /// The year, required for field sets with years (`Y`).
     pub(crate) year: Option<YearInfo>,
+    /// The month, required for field sets with months (`M`)
```

Comment:
you're documenting these fields, but not making them public.

just make them public, then you don't need the setters and the caller has full control.

Reply:
I had them public in an earlier version of this Pull Request, but then I made them private when I added the cleaner `set_date_fields` and `set_time_fields` functions. I left the docs I'd written because 🤷 .

I didn't really like them public because

1. The caller needs to know a lot about how the crate works in order to know which fields to set (having them public _in addition to_ a `set_date_fields` fn mitigates this)
2. Doesn't allow semver-changing the fields (the fields are technically public due to the GetField trait, but the trait is UnstableSealed)

Reply:
> The caller needs to know a lot about how the crate works in order to know which fields to set

Doing any kind of unchecked things requires this in the first place. I think `set_date_fields` is *too easy*.

> Doesn't allow semver-changing the fields (the fields are technically public due to the GetField trait, but the trait is UnstableSealed)

We can also mark this unstable

Reply:
> > The caller needs to know a lot about how the crate works in order to know which fields to set
> 
> Doing any kind of unchecked things requires this in the first place. I think `set_date_fields` is _too easy_.

I don't see anything wrong now or in the future of saying that `set_date_fields` is guaranteed to populate all fields required for formatting of date fields (those included under the `DateFieldSet` enum)? Seems to provide an onramp in a way that has no downsides.

> > Doesn't allow semver-changing the fields (the fields are technically public due to the GetField trait, but the trait is UnstableSealed)
> 
> We can also mark this unstable

`DateTimeInputUnstable`?

Reply:
I prefer these being private for now.